apps/src/lib/node/ledger/events/log.rs

```diff
@@ -0,0 +1,893 @@
+    /// This operation may fail if the event log has no entries.
+    ///
+    /// For as long as the caller holds an instance of an
+    /// [`EventLogSnapshot`], the state of the [`EventLog`]
+    /// at the time of calling this method can be accessed.
+    pub fn snapshot(&self) -> Result<EventLogSnapshot, Error> {
+        let log = block_in_place!(self.inner.lock.read()).unwrap();
+        log.head
+            .clone()
+            .map(|head| EventLogSnapshot {
+                head,
+                num_events: log.num_events,
+                oldest_height: log.oldest_height,
+            })
+            .ok_or(Error::EmptyLog)
+    }
+
+    /// Waits up to `deadline` for new events, and if it succeeds,
+    /// returns a new snapshot of the log.
+    pub async fn wait_snapshot(
+        &self,
+        deadline: Instant,
+    ) -> Result<EventLogSnapshot, Error> {
+        tokio::time::timeout_at(deadline, async move {
+            loop {
+                self.inner.notifier.listen().await;
+
+                if let Ok(snapshot) = self.snapshot() {
+                    break snapshot;
+                }
+            }
+        })
+        .await
+        .map_or_else(
+            // we timed out from `tokio::time::timeout_at`
+            |_| Err(Error::Timeout),
+            // we did not time out; return the snapshot
+            // of the log
+            Ok,
+        )
+    }
+
+    /// Creates a new event log.
+    fn new(params: Params) -> Self {
+        Self {
+            inner: Arc::new(EventLogInner {
+                params,
+                notifier: event_listener::Event::new(),
+                lock: RwLock::new(EventLogInnerMux {
+                    num_events: 0,
+                    oldest_height: 0.into(),
+                    head: None,
+                }),
+            }),
+        }
+    }
+
+    /// Prune the event log, ejecting old [`Event`] instances.
+    fn prune(
+        &self,
+        head: Option<Arc<LogNode>>,
+        num_events: usize,
+        height_diff: u64,
+    ) {
+        if num_events > self.inner.params.max_log_events {
+            let keep_events = calc_num_of_kept_events(num_events);
+            let snapshot = if keep_events > 0 {
+                self.prune_too_many_events(head, keep_events)
+            } else {
+                None
+            };
+            self.inner.lock.write().unwrap().install_snapshot(snapshot);
+            return;
+        }
+        if height_diff > self.inner.params.log_block_height_diff {
+            let nodes_to_keep = calc_num_of_kept_ents(height_diff);
+            let snapshot = if nodes_to_keep > 0 {
+                self.prune_old_events(head, nodes_to_keep)
+            } else {
+                None
+            };
+            self.inner.lock.write().unwrap().install_snapshot(snapshot);
+        }
+    }
+
+    /// Prune events from the log, keeping at most `max_events + excess`
```

Comment:
what is `excess`?

Reply:
ah nevermind this. forgot to clean this docstring, after I changed a part of the design.